#**Segmentação de sentenças**

NLTK -> Lib em python que permite vários processamentos de texto, entre eles a segmentação de sentenças.
- Para instalar a lib

```
pip install nltk
```
- É necessário baixar o ``` pickle ``` com os tokens
```
nltk.download('punkt')
```  



In [ ]:
!pip install nltk

In [ ]:
import nltk
nltk.download('punkt_tab')
text = "This is a English sample of the lib NLTK. NLTK is a lib for NLP."
sentence_tokenizer = nltk.data.load('tokenizers/punkt/english.pickle')
result = sentence_tokenizer.tokenize(text)
print(len(result))

for n, sentence in enumerate(result):
  print(f"{n}: {sentence}")


[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.


2
0: This is a English sample of the lib NLTK.
1: NLTK is a lib for NLP.


In [ ]:

text_pt= "Eu tenho uma gata. Essa gata dorme no meu colo! mas eu tenho outra gata que não gosta de colo"
sentence_tokenizer = nltk.data.load('tokenizers/punkt/portuguese.pickle')
result = sentence_tokenizer.tokenize(text_pt)
print(len(result))

for n, sentence in enumerate(result):
  print(f"{n}: {sentence}")


3
0: Eu tenho uma gata.
1: Essa gata dorme no meu colo!
2: mas eu tenho outra gata que não gosta de colo


# **Tokenização de texto**

In [ ]:
from nltk.tokenize import RegexpTokenizer
tokenizer = RegexpTokenizer(r'\w+')
text = "Eu gosto de café sem açúcar, mas se o café estiver forte eu ponho açúcar. Em nossa casa as pessoas costumão tomar café com bolachas"
[s.lower() for s in tokenizer.tokenize(text)]


['eu',
 'gosto',
 'de',
 'café',
 'sem',
 'açúcar',
 'mas',
 'se',
 'o',
 'café',
 'estiver',
 'forte',
 'eu',
 'ponho',
 'açúcar',
 'em',
 'nossa',
 'casa',
 'as',
 'pessoas',
 'costumão',
 'tomar',
 'café',
 'com',
 'bolachas']

# **Remoção de StopWords**

-É necessário ter uma lista de stopwords, no NLTK basta baixar essa lista:
```nltk.download('stopwords')```


In [ ]:
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [ ]:
from nltk.tokenize import RegexpTokenizer
from nltk.corpus import stopwords

tokenizer = RegexpTokenizer(r'\w+')
tokens_v0 = [s.lower() for s in tokenizer.tokenize(text)]
stop_words = stopwords.words('portuguese')
tokens_v1 = [token for token in tokens_v0 if not token in stop_words]

print(tokens_v0)
print(tokens_v1)



['eu', 'gosto', 'de', 'café', 'sem', 'açúcar', 'mas', 'se', 'o', 'café', 'estiver', 'forte', 'eu', 'ponho', 'açúcar', 'em', 'nossa', 'casa', 'as', 'pessoas', 'costumão', 'tomar', 'café', 'com', 'bolachas']
['gosto', 'café', 'açúcar', 'café', 'forte', 'ponho', 'açúcar', 'casa', 'pessoas', 'costumão', 'tomar', 'café', 'bolachas']


In [ ]:
stop_words.remove("sem")
print(stop_words)

['a', 'à', 'ao', 'aos', 'aquela', 'aquelas', 'aquele', 'aqueles', 'aquilo', 'as', 'às', 'até', 'com', 'como', 'da', 'das', 'de', 'dela', 'delas', 'dele', 'deles', 'depois', 'do', 'dos', 'e', 'é', 'ela', 'elas', 'ele', 'eles', 'em', 'entre', 'era', 'eram', 'éramos', 'essa', 'essas', 'esse', 'esses', 'esta', 'está', 'estamos', 'estão', 'estar', 'estas', 'estava', 'estavam', 'estávamos', 'este', 'esteja', 'estejam', 'estejamos', 'estes', 'esteve', 'estive', 'estivemos', 'estiver', 'estivera', 'estiveram', 'estivéramos', 'estiverem', 'estivermos', 'estivesse', 'estivessem', 'estivéssemos', 'estou', 'eu', 'foi', 'fomos', 'for', 'fora', 'foram', 'fôramos', 'forem', 'formos', 'fosse', 'fossem', 'fôssemos', 'fui', 'há', 'haja', 'hajam', 'hajamos', 'hão', 'havemos', 'haver', 'hei', 'houve', 'houvemos', 'houver', 'houvera', 'houverá', 'houveram', 'houvéramos', 'houverão', 'houverei', 'houverem', 'houveremos', 'houveria', 'houveriam', 'houveríamos', 'houvermos', 'houvesse', 'houvessem', 'houvésse

In [ ]:
tokenizer = RegexpTokenizer(r'\w+')
tokens_v0 = [s.lower() for s in tokenizer.tokenize(text)]
# stop_words = stopwords.words('portuguese')
tokens_v1 = [token for token in tokens_v0 if not token in stop_words]

print(tokens_v0)
print(tokens_v1)

['eu', 'gosto', 'de', 'café', 'sem', 'açúcar', 'mas', 'se', 'o', 'café', 'estiver', 'forte', 'eu', 'ponho', 'açúcar', 'em', 'nossa', 'casa', 'as', 'pessoas', 'costumão', 'tomar', 'café', 'com', 'bolachas']
['gosto', 'café', 'sem', 'açúcar', 'café', 'forte', 'ponho', 'açúcar', 'casa', 'pessoas', 'costumão', 'tomar', 'café', 'bolachas']


# **Lemmatization**

In [ ]:
nltk.download('wordnet')
from nltk.test.portuguese_en_fixt import setup_module


[nltk_data] Downloading package wordnet to /root/nltk_data...


In [ ]:
from nltk.stem import WordNetLemmatizer
wordnet_lemmatizer = WordNetLemmatizer()
lemma_words = [wordnet_lemmatizer.lemmatize(word) for word in tokens_v0]
print(tokens_v0)
print(lemma_words)


['eu', 'gosto', 'de', 'café', 'sem', 'açúcar', 'mas', 'se', 'o', 'café', 'estiver', 'forte', 'eu', 'ponho', 'açúcar', 'em', 'nossa', 'casa', 'as', 'pessoas', 'costumão', 'tomar', 'café', 'com', 'bolachas']
['eu', 'gosto', 'de', 'café', 'sem', 'açúcar', 'ma', 'se', 'o', 'café', 'estiver', 'forte', 'eu', 'ponho', 'açúcar', 'em', 'nossa', 'casa', 'a', 'pessoas', 'costumão', 'tomar', 'café', 'com', 'bolachas']


# **Stemming**

In [ ]:
nltk.download('rslp')
from nltk.stem.porter import PorterStemmer
from nltk.stem import RSLPStemmer
rslp_stemmer = RSLPStemmer()
porter_stemmer = PorterStemmer()
stem_words = [porter_stemmer.stem(word) for word in tokens_v0]
stem_words2 = [rslp_stemmer.stem(word) for word in tokens_v0]

print(tokens_v0)
print(stem_words)
print(stem_words2)

['eu', 'gosto', 'de', 'café', 'sem', 'açúcar', 'mas', 'se', 'o', 'café', 'estiver', 'forte', 'eu', 'ponho', 'açúcar', 'em', 'nossa', 'casa', 'as', 'pessoas', 'costumão', 'tomar', 'café', 'com', 'bolachas']
['eu', 'gosto', 'de', 'café', 'sem', 'açúcar', 'ma', 'se', 'o', 'café', 'estiv', 'fort', 'eu', 'ponho', 'açúcar', 'em', 'nossa', 'casa', 'as', 'pessoa', 'costumão', 'tomar', 'café', 'com', 'bolacha']
['eu', 'gost', 'de', 'café', 'sem', 'açúc', 'mas', 'se', 'o', 'café', 'estiv', 'fort', 'eu', 'ponh', 'açúc', 'em', 'noss', 'cas', 'as', 'pesso', 'costum', 'tom', 'café', 'com', 'bolach']


[nltk_data] Downloading package rslp to /root/nltk_data...
[nltk_data]   Unzipping stemmers/rslp.zip.


## Stemming Vs Lemmatization

In [ ]:
words_sample= ['amigos', 'amigas', 'amizade', 'carreira', 'carreiras', 'tempestade']
lemma_words = [wordnet_lemmatizer.lemmatize(word) for word in words_sample]
stem_words = [porter_stemmer.stem(word) for word in words_sample]

print(words_sample)
print(lemma_words)
print(stem_words)


['amigos', 'amigas', 'amizade', 'carreira', 'carreiras', 'tempestade']
['amigo', 'amigas', 'amizade', 'carreira', 'carreiras', 'tempestade']
['amigo', 'amiga', 'amizad', 'carreira', 'carreira', 'tempestad']


# Extração de caracteísticas

## Part-of-speech

Em inglês:
1. CC - Coordinating conjunction
2. CD - Cardinal number
3. DT - Determiner
4. EX - Existential there
5. FW - Foreign word
6. IN - Preposition or subordinating conjunction
7. JJ - Adjective
8. JJR - Adjective, comparative
9. JJS - Adjective, superlative
10. LS - List item marker
11. MD - Modal
12. NN - Noun, singular or mass
13. NNS - Noun, plural
14. NNP - Proper noun, singular
15. NNPS - Proper noun, plural
16. PDT - Predeterminer
17. POS - Possessive ending
18. PRP - Personal pronoun
19. PRP - 	Possessive pronoun
20. RB - Adverb
21. RBR - Adverb, comparative
22. RBS - Adverb, superlative
23. RP - Particle
24. SYM - Symbol
25. TO - to
26. UH - Interjection
27. VB - Verb, base form
28. VBD - Verb, past tense
29. VBG - Verb, gerund or present participle
30. VBN - Verb, past participle
31. VBP - Verb, non-3rd person singular present
32. VBZ - Verb, 3rd person singular present
33. WDT - Wh-determiner
34. WP - Wh-pronoun
35. WP - 	Possessive wh-pronoun
36. WRB - Wh-adverb

In [ ]:
import nltk
nltk.download('averaged_perceptron_tagger_eng')
text ="Happiness can be found, even in the darkest of times, if one only remembers to turn on the light."
tokens = [s.lower() for s in tokenizer.tokenize(text)]
nltk.pos_tag(tokens)


[nltk_data] Downloading package averaged_perceptron_tagger_eng to
[nltk_data]     /root/nltk_data...
[nltk_data]   Unzipping taggers/averaged_perceptron_tagger_eng.zip.


[('happiness', 'NN'),
 ('can', 'MD'),
 ('be', 'VB'),
 ('found', 'VBN'),
 ('even', 'RB'),
 ('in', 'IN'),
 ('the', 'DT'),
 ('darkest', 'NN'),
 ('of', 'IN'),
 ('times', 'NNS'),
 ('if', 'IN'),
 ('one', 'CD'),
 ('only', 'RB'),
 ('remembers', 'VBZ'),
 ('to', 'TO'),
 ('turn', 'VB'),
 ('on', 'IN'),
 ('the', 'DT'),
 ('light', 'NN')]

## Extração de características TF


In [ ]:
print(tokens_v0)
print(tokens_v1)

['eu', 'gosto', 'de', 'café', 'sem', 'açúcar', 'mas', 'se', 'o', 'café', 'estiver', 'forte', 'eu', 'ponho', 'açúcar', 'em', 'nossa', 'casa', 'as', 'pessoas', 'costumão', 'tomar', 'café', 'com', 'bolachas']
['gosto', 'café', 'sem', 'açúcar', 'café', 'forte', 'ponho', 'açúcar', 'casa', 'pessoas', 'costumão', 'tomar', 'café', 'bolachas']


In [ ]:
import pandas as pd
import numpy as np

df = pd.DataFrame(data=[np.zeros(11)], columns=list(set(tokens_v1)), index=["N","TF"])
quant_token = len(list(set(tokens_v1)))
print(df)
print(quant_token)
for word in tokens_v0:
  if word in df.keys():
    df[word]["N"] += 1

for token in df.keys():
  df[token]["TF"] = df[token]["N"]/quant_token

df


    tomar  casa  açúcar  gosto  bolachas  café  ponho  pessoas  costumão  sem  \
N     0.0   0.0     0.0    0.0       0.0   0.0    0.0      0.0       0.0  0.0   
TF    0.0   0.0     0.0    0.0       0.0   0.0    0.0      0.0       0.0  0.0   

    forte  
N     0.0  
TF    0.0  
11


<ipython-input-91-f29cc538305a>:10: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df[word]["N"] += 1
<ipython-input-91-f29cc538305a>:13: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values throu

,tomar,casa,açúcar,gosto,bolachas,café,ponho,pessoas,costumão,sem,forte
N,1.000000,1.000000,2.000000,1.000000,1.000000,3.000000,1.000000,1.000000,1.000000,1.000000,1.000000
TF,0.090909,0.090909,0.181818,0.090909,0.090909,0.272727,0.090909,0.090909,0.090909,0.090909,0.090909


# Pratica:
## Pratica em grupo, 1 integrante do grupo deve enviar esse notebook com o nome completo de cada integrante do grupo na célula a baixo. Integrante do grupo que não tenha o nome na célula não receberá nota.
1. Rode o código da célula a baixo para gerar o texto base
2. (20ps) Faça uma primeira exploração dos dados. Crie uma célula de texto descrevendo os dados. ex: Do que são esses dados? Que tipo de "sujeiras" tem nele? De que ano são? etc.
3. (30ps) Limpe os dados e os separe em sentenças
4. (20ps) Pra cada sentença faça:
  - Tokenização
  - Remoção de stopwords usando a lista da lib NLTK (reflita se tem alguma palavra que você colocaria/removeria da lista de stop words, qual seria?)
5. (20ps) Por fim, compare a Lemmatization com Stemming de cada frase.


# Nomes dos integrantes do grupo em ordem alfabética:

---
* Arthur Dantas - altds@cesar.school
* Gabriel Ximenes - gxbc@cesar.school
* Geronimo Morais - gmln@cesar.school
* Lucas Barbosa - lbs3@cesar.school
* Maxmiliano Augusto - mags@cesar.school




## 01.Gerando texto base

In [ ]:
import nltk
nltk.download('mac_morpho')
sents = nltk.corpus.mac_morpho.sents()
text = "<p>"
for s in sents[:10]:
  text += " ".join(s)
  text += ".</p> |<p>"

print(len(text))
print(text[:1000])


[nltk_data] Downloading package mac_morpho to /root/nltk_data...


1157
<p>Jersei atinge média de Cr$ 1,4 milhão em a venda de a Pinhal em São Paulo.</p> |<p>Programe sua viagem a a Exposição Nacional do Zebu , que começa dia 25.</p> |<p>Safra recorde e disponibilidade de crédito ativam vendas de máquinas agrícolas.</p> |<p>A degradação de as terras por o mau uso de os solos avança em o.</p> |<p>A desertificação tornou crítica a produtividade de 52 mil km² em a região.</p> |<p>Em o dia 15 , Dia da Conservação do Solo , o único fato a festejar pode ser a Convenção Internacional sobre Desertificação.</p> |<p>A produção brasileira de pintos de corte totalizou , em fevereiro último , 166 milhões , volume 6,79 % superior a o registrado em fevereiro de 93 , segundo dados de a Associação Brasileira dos Produtores de Pinto de Corte ( Apinco ).</p> |<p>A Apinco destaca em seu boletim mensal que o setor avícola está otimista com o atual programa de estabilização econômica.</p> |<p>" A melhoria de o padrão aquisitivo resulta em maior demanda de carne de frango "

[nltk_data]   Unzipping corpora/mac_morpho.zip.


## 02.Identificando sujeiras
- O documento se trata de um site de notícia, nítido pelas tags HTML. O conteúdo abroda principalmente assuntos de agropecuária. Podemos confimrar isso pela presença das palavras "solo", "safra", "avícola" .

- Há indícios que o texto seja do ano de 1994. A evidência para isto, consta na sentença "A produção brasileira de pintos de corte totalizou  , em fevereiro último  , 166 milhões  , volume 6 ,79 % superior a o registrado em fevereiro de 93". Como a sentença fala do último mês de fevereiro em comparação com o mesmo mês do ano de 1993, podemos supor que o texto está no presente - logo, 1994.

- Há muitas tags HTML que precisariam ser removidas para melhor análise das palavras(tokens). Elas podem ser consideradas sujeiras.

In [ ]:
text

'<p>Jersei atinge média de Cr$ 1,4 milhão em a venda de a Pinhal em São Paulo.</p> |<p>Programe sua viagem a a Exposição Nacional do Zebu , que começa dia 25.</p> |<p>Safra recorde e disponibilidade de crédito ativam vendas de máquinas agrícolas.</p> |<p>A degradação de as terras por o mau uso de os solos avança em o.</p> |<p>A desertificação tornou crítica a produtividade de 52 mil km² em a região.</p> |<p>Em o dia 15 , Dia da Conservação do Solo , o único fato a festejar pode ser a Convenção Internacional sobre Desertificação.</p> |<p>A produção brasileira de pintos de corte totalizou , em fevereiro último , 166 milhões , volume 6,79 % superior a o registrado em fevereiro de 93 , segundo dados de a Associação Brasileira dos Produtores de Pinto de Corte ( Apinco ).</p> |<p>A Apinco destaca em seu boletim mensal que o setor avícola está otimista com o atual programa de estabilização econômica.</p> |<p>" A melhoria de o padrão aquisitivo resulta em maior demanda de carne de frango ".</p

## 03.Limpando os dados e separando em sentenças

In [ ]:
text = text.replace("<p>", "").replace("</p>", "").replace("|", "").replace(",", " ,").replace("!", " !").replace("?", " ?").replace("'", "")
text

'Jersei atinge média de Cr$ 1 ,4 milhão em a venda de a Pinhal em São Paulo. Programe sua viagem a a Exposição Nacional do Zebu  , que começa dia 25. Safra recorde e disponibilidade de crédito ativam vendas de máquinas agrícolas. A degradação de as terras por o mau uso de os solos avança em o. A desertificação tornou crítica a produtividade de 52 mil km² em a região. Em o dia 15  , Dia da Conservação do Solo  , o único fato a festejar pode ser a Convenção Internacional sobre Desertificação. A produção brasileira de pintos de corte totalizou  , em fevereiro último  , 166 milhões  , volume 6 ,79 % superior a o registrado em fevereiro de 93  , segundo dados de a Associação Brasileira dos Produtores de Pinto de Corte ( Apinco ). A Apinco destaca em seu boletim mensal que o setor avícola está otimista com o atual programa de estabilização econômica. " A melhoria de o padrão aquisitivo resulta em maior demanda de carne de frango ". O secretário de a agricultura paulista  , Roberto Rodrigues 

In [ ]:
result = sentence_tokenizer.tokenize(text)
list_sentences = []
for sentence in result:
  list_sentences.append(sentence)


In [ ]:
list_sentences

['Jersei atinge média de Cr$ 1 ,4 milhão em a venda de a Pinhal em São Paulo.',
 'Programe sua viagem a a Exposição Nacional do Zebu  , que começa dia 25.',
 'Safra recorde e disponibilidade de crédito ativam vendas de máquinas agrícolas.',
 'A degradação de as terras por o mau uso de os solos avança em o.',
 'A desertificação tornou crítica a produtividade de 52 mil km² em a região.',
 'Em o dia 15  , Dia da Conservação do Solo  , o único fato a festejar pode ser a Convenção Internacional sobre Desertificação.',
 'A produção brasileira de pintos de corte totalizou  , em fevereiro último  , 166 milhões  , volume 6 ,79 % superior a o registrado em fevereiro de 93  , segundo dados de a Associação Brasileira dos Produtores de Pinto de Corte ( Apinco ).',
 'A Apinco destaca em seu boletim mensal que o setor avícola está otimista com o atual programa de estabilização econômica. "',
 'A melhoria de o padrão aquisitivo resulta em maior demanda de carne de frango ".',
 'O secretário de a agric

## 04.Tokenização

In [ ]:
from nltk.corpus import stopwords

tokenizer = RegexpTokenizer(r'\w+')
stop_words = stopwords.words('portuguese')

Obs: Decidimos optar pela remoção dos números, pois ficavam muito soltos dentro do contexto!!

Além disso decidimos tratar as sentenças separadamente, aplicando o set separadamente para cada frase para que não se perdessem palavras essenciais para cada contexto.

In [ ]:
import re
tokens = []
pre_list = []
dict = {}
#stop_words.remove("são")
for n, sentence in enumerate(list_sentences):
  dict[n] = sentence
for key,value in dict.items():
    pre_list += value.lower().replace(",", "").split(" ")
    frase = [dict.fromkeys([s.lower() for s in tokenizer.tokenize(value)])]  # Removendo redundância nas palavras porém mantendo a ordem da frase
    for v in frase:
        v = [token for token in v if not token in stop_words]  # Removendo Stopwords
        tokens += v
        for n in tokens:
          if n.isdigit():
            tokens.remove(n)
print(pre_list)
print(tokens)
print(f"Total de palavras depois do Set & StopWords:{len(pre_list)}")
print(f"Total de palavras depois do Set & StopWords:{len(tokens)}")

['jersei', 'atinge', 'média', 'de', 'cr$', '1', '4', 'milhão', 'em', 'a', 'venda', 'de', 'a', 'pinhal', 'em', 'são', 'paulo.', 'programe', 'sua', 'viagem', 'a', 'a', 'exposição', 'nacional', 'do', 'zebu', '', '', 'que', 'começa', 'dia', '25.', 'safra', 'recorde', 'e', 'disponibilidade', 'de', 'crédito', 'ativam', 'vendas', 'de', 'máquinas', 'agrícolas.', 'a', 'degradação', 'de', 'as', 'terras', 'por', 'o', 'mau', 'uso', 'de', 'os', 'solos', 'avança', 'em', 'o.', 'a', 'desertificação', 'tornou', 'crítica', 'a', 'produtividade', 'de', '52', 'mil', 'km²', 'em', 'a', 'região.', 'em', 'o', 'dia', '15', '', '', 'dia', 'da', 'conservação', 'do', 'solo', '', '', 'o', 'único', 'fato', 'a', 'festejar', 'pode', 'ser', 'a', 'convenção', 'internacional', 'sobre', 'desertificação.', 'a', 'produção', 'brasileira', 'de', 'pintos', 'de', 'corte', 'totalizou', '', '', 'em', 'fevereiro', 'último', '', '', '166', 'milhões', '', '', 'volume', '6', '79', '%', 'superior', 'a', 'o', 'registrado', 'em', 'fever

# 05. Lemmatization x Stemming

In [ ]:
lemma_words = [wordnet_lemmatizer.lemmatize(word) for word in tokens]
stem_words = [porter_stemmer.stem(word) for word in tokens]


print(f"Quantidade de palavras usando Lemmatization:{len(lemma_words)}")
print(f"Quantidade de palavras usando Stemming:{len(stem_words)}")

print(lemma_words)
print(stem_words)

print("---Lemmatization---")
for n, value in enumerate(tokens):
    if value == lemma_words[n]:
        pass
    else:
        print(f"A palavra '{value}' foi alterada para '{lemma_words[n]}'")

print("---Stemming---")
for n, value in enumerate(tokens):
    if value == stem_words[n]:
        pass
    else:
        print(f"A palavra '{value}' foi alterada para '{stem_words[n]}'")


Quantidade de palavras usando Lemmatization:97
Quantidade de palavras usando Stemming:97
['jersei', 'atinge', 'média', 'cr', 'milhão', 'venda', 'pinhal', 'são', 'paulo', 'programe', 'viagem', 'exposição', 'nacional', 'zebu', 'começa', 'dia', 'safra', 'recorde', 'disponibilidade', 'crédito', 'ativam', 'vendas', 'máquinas', 'agrícolas', 'degradação', 'terras', 'mau', 'uso', 'solo', 'avança', 'desertificação', 'tornou', 'crítica', 'produtividade', 'mil', 'km²', 'região', 'dia', 'conservação', 'solo', 'único', 'fato', 'festejar', 'pode', 'convenção', 'internacional', 'sobre', 'desertificação', 'produção', 'brasileira', 'pinto', 'corte', 'totalizou', 'fevereiro', 'último', 'milhões', 'volume', 'superior', 'registrado', 'segundo', 'dado', 'associação', 'produtores', 'pinto', 'apinco', 'apinco', 'destaca', 'boletim', 'mensal', 'setor', 'avícola', 'otimista', 'atual', 'programa', 'estabilização', 'econômica', 'melhoria', 'padrão', 'aquisitivo', 'resulta', 'maior', 'demanda', 'carne', 'frango',

## Calculando o TF

In [ ]:
df = pd.DataFrame(data=[np.zeros(len(list(set(tokens))))], columns=list(set(tokens)), index=["N","TF"])
quant_token = len(list(set(tokens)))

In [ ]:
for word in pre_list:
  if word in df.keys():
    df[word]["N"] += 1

for token in df.keys():
  df[token]["TF"] = df[token]["N"]/quant_token

df

<ipython-input-101-acc9a0d1bab3>:3: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df[word]["N"] += 1
<ipython-input-101-acc9a0d1bab3>:6: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values throu

,anunciado,ativam,tornou,atual,disponibilidade,jersei,crédito,festejar,pintos,degradação,...,pinhal,apinco,nacional,dados,vendas,secretário,volume,milhão,produção,região
N,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,...,1.000000,2.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,0.0
TF,0.010638,0.010638,0.010638,0.010638,0.010638,0.010638,0.010638,0.010638,0.010638,0.010638,...,0.010638,0.021277,0.010638,0.010638,0.010638,0.010638,0.010638,0.010638,0.010638,0.0
